### ISSUE TITLE
Add Tensor.requires_grad & optional grad calculation

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#601]
Currently there are a couple of spots where the tensors won't ever need gradients, but currently dfdx does calculate them. The spots are:

1. The very first input to the network doesn't need gradients (Related to #231)
2. The max in logsumexp used for stabliziation does not need gradients. This applies to both the sub & the add.

Particularly for logsumexp, the backward operation for sub(max) and add(max) would need to launch 1 less kernel for cuda, and do half the work for CPU.

**Comment by coreylowman** [comment]
Closing this for now. For logsumexp this can be resolved another way with tape manipulation. And the #231 can still represent this work
